CMakeLists.txt

```diff
@@ -3,17 +3,26 @@ project(GPUTreeShap LANGUAGES CXX CUDA)
 set(CMAKE_CXX_STANDARD 14)
 set(CMAKE_CUDA_STANDARD 14)
 
-find_package(GTest)
-include_directories(${CMAKE_SOURCE_DIR})
-add_executable(TestGPUTreeShap tests/test_gpu_treeshap.cu GPUTreeShap/gpu_treeshap.h)
-target_include_directories(TestGPUTreeShap  PRIVATE ${GTEST_INCLUDE_DIRS})
-if(MSVC)
+option(BUILD_GTEST "Build google tests" OFF)
+
+# GPUTreeSHAP target is header-only
+add_library(GPUTreeShap INTERFACE)
+add_library(GPUTreeShap::GPUTreeShap ALIAS GPUTreeShap)
+target_sources(GPUTreeShap INTERFACE ${GPUTreeShap_SOURCE_DIR}/GPUTreeShap/gpu_treeshap.h)
+target_include_directories(GPUTreeShap INTERFACE ${GPUTreeShap_SOURCE_DIR})
+
+if(BUILD_GTEST)
+  find_package(GTest REQUIRED)
+  add_executable(TestGPUTreeShap tests/test_gpu_treeshap.cu)
+  target_link_libraries(TestGPUTreeShap PRIVATE GPUTreeShap)
+  if(MSVC)
     target_compile_options(TestGPUTreeShap PRIVATE $<$<COMPILE_LANGUAGE:CUDA>:-Xcompiler /WX>)
-else()
+  else()
     target_compile_options(TestGPUTreeShap PRIVATE $<$<COMPILE_LANGUAGE:CUDA>:-Xcompiler -Werror,-Wall,-Wextra>)
+  endif()
+  target_link_libraries(TestGPUTreeShap  PRIVATE GTest::GTest GTest::Main)
+  add_custom_command(TARGET TestGPUTreeShap PRE_BUILD COMMAND cpplint ${GPUTreeShap_SOURCE_DIR}/GPUTreeShap/gpu_treeshap.h ${GPUTreeShap_SOURCE_DIR}/tests/test_gpu_treeshap.cu ${GPUTreeShap_SOURCE_DIR}/example/example.cu)
 endif()
-target_link_libraries(TestGPUTreeShap  PRIVATE  ${GTEST_BOTH_LIBRARIES})
-
-add_custom_command(TARGET TestGPUTreeShap PRE_BUILD COMMAND cpplint ${CMAKE_SOURCE_DIR}/GPUTreeShap/gpu_treeshap.h ${CMAKE_SOURCE_DIR}/tests/test_gpu_treeshap.cu ${CMAKE_SOURCE_DIR}/example/example.cu)
 
-add_executable(GPUTreeShapExample example/example.cu GPUTreeShap/gpu_treeshap.h)
+add_executable(GPUTreeShapExample example/example.cu)
```

Comment:
Building the example could be optional, so it doesn't get built by the parent project.

Reply:
Also it's useful to include headers in add_executable so visual studio includes them in projects.

Reply:
Sure, I'll make it optional.

Reply:
> Also it's useful to include headers in add_executable so visual studio includes them in projects.

Let me try building it in MSVC.

Reply:
@RAMitchell The header appears under the parent project as follows:

![Screen Shot 2020-08-24 at 7 39 21 PM](https://user-images.githubusercontent.com/2532981/91116594-8c570f80-e641-11ea-95eb-83f9f1fe9340.png)

This is because of the `target_sources()` command, which exports the header to all consumer targets. (See [this link](https://stackoverflow.com/a/29218394) for reference.)
```cmake
target_sources(GPUTreeShap INTERFACE ${GPUTreeShap_SOURCE_DIR}/GPUTreeShap/gpu_treeshap.h)
```